packages/relay-runtime/store/RelayPublishQueue.js

```diff
@@ -125,32 +122,45 @@ class RelayPublishQueue {
    */
   revertAll(): void {
     this._pendingBackupRebase = true;
-    this._pendingOptimisticUpdates.clear();
-    this._appliedOptimisticUpdates.clear();
+    this._pendingUpdates = this._pendingUpdates.filter(
+      update => update.kind !== 'optimistic',
+    );
   }
 
   /**
    * Schedule applying a payload to the store on the next `run()`.
+   * If provided, this will revert the corresponding optimistic update
    */
   commitPayload(
     operation: OperationSelector,
     {fieldPayloads, source}: RelayResponsePayload,
     updater?: ?SelectorStoreUpdater,
+    optimisticUpdate?: OptimisticUpdate,
   ): void {
-    this._pendingBackupRebase = true;
-    this._pendingData.add({
+    const serverData = {
       kind: 'payload',
       payload: {fieldPayloads, operation, source, updater},
-    });
+    };
+    if (optimisticUpdate) {
+      const updateIdx = findUpdaterIdx(this._pendingUpdates, optimisticUpdate);
+      if (updateIdx !== -1) {
+        this._pendingBackupRebase = true;
```

Comment:
note that before with arbitrary transform application order, it had to rebase each time a commit happened (from server, source, or local). now, we only rebase if it comes in & is replacing an optimistic update. net win for performance.